infrastructure/src/main/java/org/corfudb/infrastructure/CustomServerRouter.java

```diff
@@ -0,0 +1,186 @@
+        } catch (Exception e) {
+            log.error("Fatal error: Failed to create serverAdapter", e);
+            throw new UnrecoverableCorfuError(e);
+        }
+    }
+
+    // ============ IServerRouter Methods =============
+
+    @Override
+    public void sendResponse(ChannelHandlerContext ctx, CorfuMsg inMsg, CorfuMsg outMsg) {
+        log.info("Ready to send response {}", outMsg.getMsgType());
+        outMsg.copyBaseFields(inMsg);
+        try {
+            serverAdapter.send(CorfuMessageConverter.toProtoBuf(outMsg));
+            log.trace("Sent response: {}", outMsg);
+        } catch (IllegalArgumentException e) {
+            log.warn("Illegal response type. Ignoring message.", e);
+        }
+    }
+
+    @Override
+    public Optional<Layout> getCurrentLayout() {
+        return Optional.empty();
+    }
+
+    @Override
+    public List<AbstractServer> getServers() {
+        return servers;
+    }
+
+    @Override
+    public void setServerContext(ServerContext serverContext) {
+
+    }
+
+    /**
+     * Receive messages from the 'custom' serverAdapter implementation. This message will be forwarded
+     * for processing.
+     *
+     * @param protoMessage
+     */
+    public void receive(CorfuMessage protoMessage) {
+        CorfuMsg corfuMsg;
+        try {
+            log.info("Received message {}", protoMessage.getType().name());
+            // Transform protoBuf into CorfuMessage
+            corfuMsg = CorfuMessageConverter.fromProtoBuf(protoMessage);
+        } catch (CorfuMessageProtoBufException e) {
+            log.error("Exception while trying to convert {} from protoBuf", protoMessage.getType(), e);
+            return;
+        }
+
+        AbstractServer handler = handlerMap.get(corfuMsg.getMsgType());
+        if (handler == null) {
+            // The message was unregistered, we are dropping it.
+            log.warn("Received unregistered message {}, dropping", corfuMsg);
+        } else {
+            if (validateEpoch(corfuMsg)) {
+                // Route the message to the handler.
+                if (log.isTraceEnabled()) {
+                    log.trace("Message routed to {}: {}", handler.getClass().getSimpleName(), corfuMsg);
+                }
+
+                try {
+                    handler.handleMessage(corfuMsg, null, this);
+                } catch (Throwable t) {
+                    log.error("channelRead: Handling {} failed due to {}:{}",
+                            corfuMsg != null ? corfuMsg.getMsgType() : "UNKNOWN",
+                            t.getClass().getSimpleName(),
+                            t.getMessage(),
+                            t);
+                }
+            }
+        }
+    }
+
+    /**
+     * Validate the epoch of a CorfuMsg, and send a WRONG_EPOCH response if
+     * the server is in the wrong epoch. Ignored if the message type is reset (which
```

Comment:
The comment does not match the check here.. the actual check looks for the ignoreEpoch property of the message and not the reset type.  Can we change at 1 place to make it consistent?

Also, is it safe to ignore the epoch for reset messages?  What if a reset message from an older epoch is delivered late and should not be honoured?